client/src/pages/ProductDetails.tsx

```diff
@@ -0,0 +1,88 @@
+import { ReactElement, useState, useEffect } from 'react';
+import { fetchProduct, Product, toDollars } from '../lib';
+import { useParams } from 'react-router-dom';
+import { Tab } from '@headlessui/react';
+
+export default function ProductDetails(): ReactElement {
+  const { productId } = useParams();
+  const [product, setProduct] = useState<Product>();
+  const [isLoading, setIsLoading] = useState(true);
+  const [error, setError] = useState<unknown>();
+
+  useEffect(() => {
+    async function loadProduct(productId: number) {
+      try {
+        const product = await fetchProduct(productId);
+        setProduct(product);
+      } catch (err) {
+        setError(err);
+      } finally {
+        setIsLoading(false);
+      }
+    }
+    if (productId) {
+      setIsLoading(true);
+      loadProduct(+productId);
+    }
+  }, [productId]);
+
+  if (isLoading) return <div>Loading...</div>;
+  if (error)
+    return (
+      <div>
+        Error Loading Product {productId}:{' '}
+        {error instanceof Error ? error.message : 'Unknown Error'}
+      </div>
+    );
+  if (!product) return <div>Product not found</div>;
+
+  return (
+    <div className="bg-white">
+      <div className="mx-auto max-w-2xl px-4 py-16 sm:px-6 sm:py-24 lg:max-w-7xl lg:px-8">
+        <div className="lg:grid lg:grid-cols-2 lg:items-start lg:gap-x-8">
+          {/* Image gallery */}
+          <Tab.Group as="div" className="flex flex-col-reverse">
+            {/* Image selector */}
```

Comment:
A comment like this usually suggests that it's time to break it out into a separate component.